src/control/probe.rs

```diff
@@ -0,0 +1,469 @@
+    index: usize,
+
+    /// Iterator over full buckets in the group.
+    iter: BitMaskIter,
+
+    /// Marker indicating that we have items of type T.
+    marker: PhantomData<*const T>,
+}
+impl<T> Clone for ProbeItems<T> {
+    fn clone(&self) -> Self {
+        ProbeItems {
+            probe: self.probe.clone(),
+            iter: self.iter.clone(),
+            index: self.index,
+            marker: PhantomData,
+        }
+    }
+}
+impl<T> Iterator for ProbeItems<T> {
+    type Item = (usize, NonNull<T>);
+
+    fn next(&mut self) -> Option<(usize, NonNull<T>)> {
+        loop {
+            if let Some(bit) = self.iter.next() {
+                // SAFETY: This function was made for us, and we uphold our own constraints.
+                return Some(unsafe {
+                    group_probe_items(self.probe.ctrl, self.index, bit, self.probe.bucket_mask)
+                });
+            } else if self.probe.current().has_empty() {
+                return None;
+            } else {
+                // SAFETY: The probe always returns an item.
+                let group_probe = unsafe { self.probe.next().unwrap_unchecked() };
+
+                // SAFETY: The hash table contains this type of items.
+                let full = unsafe { group_probe.full::<T>() };
+                self.iter = full.iter;
+                self.index = full.index;
+            }
+        }
+    }
+}
+
+/// Probe which iterates through groups in the control data to search for
+/// buckets for a given hash value.
+///
+/// Sequence based on triangular numbers, which is guaranteed (since our table
+/// size is a power of two) to visit every group of elements exactly once.
+///
+/// A triangular probe has us jump by 1 more group every time. So first we
+/// jump by 1 group (meaning we just continue our linear scan), then 2 groups
+/// (skipping over 1 group), then 3 groups (skipping over 2 groups), and so on.
+///
+/// Proof that the probe will visit every group in the table:
+/// <https://fgiesen.wordpress.com/2015/02/22/triangular-numbers-mod-2n/>
+#[derive(Clone)]
+pub(crate) struct Probe {
+    /// Current position in sequence.
+    pos: usize,
+
+    /// Offset from current position to next group.
+    stride: usize,
+
+    /// Pointer to control data.
+    ctrl: NonNull<Tag>,
+
+    /// Number of buckets, minus one.
+    bucket_mask: usize,
+
+    /// Tag containing the hash of the item.
+    hash_tag: Tag,
+
+    /// Initial hash, for debugging.
+    #[cfg(debug_assertions)]
+    hash: u64,
+
+    /// Current iteration, for debugging.
+    #[cfg(debug_assertions)]
+    iter: usize,
+}
+impl Probe {
+    /// Creates a new probe for a table for a given hash.
+    ///
+    /// # Safety
+    ///
+    /// The given bucket mask must be equal to the power-of-two number of buckets minus one,
+    /// and the given control pointer must be a valid hash table.
+    pub(crate) unsafe fn new(ctrl: NonNull<Tag>, bucket_mask: usize, hash: u64) -> Probe {
```

Comment:
All non-generic functions need to have `#[inline]`. And hot generic functions should as well, since it causes them to be replicated in all codegen units they are referenced by instead of relying on LTO to inline them.